File: Java/EX5.java
import java.util.ArrayList;
import java.util.List;

// Classe base Animal
abstract class Animal2 {
    // Método abstrato para ser implementado pelas subclasses
    public abstract String som();
}

// Classe derivada Cachorro
class Cachorro2 extends Animal {
    @Override
    public String som() {
        return "Au Au";
    }
}

// Classe derivada Gato
class Gato2 extends Animal {
    @Override
    public String som() {
        return "Miau";
    }
}

// Classe principal
class Main1 {
    // Função que recebe uma lista de objetos Animal
    public static void emitirSons(List<Animal> animais) {
        for (Animal animal : animais) {
            System.out.println(animal.som());
        }
    }

    public static void main(String[] args) {
        // Criando instâncias das classes
        Cachorro cachorro1 = new Cachorro();
        Gato gato1 = new Gato();
        Cachorro cachorro2 = new Cachorro();

        // Criando uma lista de animais
        List<Animal> animais = new ArrayList<>();
        animais.add(cachorro1);
        animais.add(gato1);
        animais.add(cachorro2);

        // Chamando a função que emite os sons
        emitirSons(animais);
    }
}

File: Go/EX5.go
package main

import "fmt"

// Interface Animal
type Animal interface {
	Som() string
}

// Struct Cachorro
type Cachorro struct{}

// Implementação do método Som para Cachorro
func (c Cachorro) Som() string {
	return "Au Au"
}

// Struct Gato
type Gato struct{}

// Implementação do método Som para Gato
func (g Gato) Som() string {
	return "Miau"
}

// Função que recebe uma lista de objetos Animal
func emitirSons(animais []Animal) {
	for _, animal := range animais {
		fmt.Println(animal.Som())
	}
}

// Função principal
func main() {
	// Criando instâncias das structs
	cachorro1 := Cachorro{}
	gato1 := Gato{}
	cachorro2 := Cachorro{}

	// Criando uma lista de animais
	animais := []Animal{cachorro1, gato1, cachorro2}

	// Chamando a função que emite os sons
	emitirSons(animais)
}
